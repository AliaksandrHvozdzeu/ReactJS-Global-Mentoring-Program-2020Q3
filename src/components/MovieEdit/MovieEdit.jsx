import React from 'react';
import PropTypes from 'prop-types';
import Input from '../Input';
import MultiSelect from '../MultiSelectInput';
import Constants from '../../constants';
import CloseButton from '../CloseButton';
import Calendar from '../Calendar';
import './style.css';

const handleSubmit = () => {

};

const validate = () => {

};

const multiselectValidate = (data) => {

};

const reset = () => {

};

const submit = () => {

};

const MovieEdit = ({ modalTitle, closeAction, details }) => (
  <>
    <CloseButton closeAction={closeAction}/>
    <form className="add-movie" onSubmit={handleSubmit}>
      <p className="modal-title">{modalTitle}</p>
      <Input id="movie-id"
             title="Movie id"
             visibility=""
             value={details.id}
             onChange={validate}/>
      <Input id="movie-title"
             title="Title"
             visibility=""
             value={details.title}
             onChange={validate}/>
      <Calendar id="movie-release"
                title="Release date"
                value={details.release}
                onChange={validate}/>
      <Input id="movie-url"
             title="Movie url"
             visibility=""
             value={details.poster}
             onChange={validate}/>
      <MultiSelect title="genre"
                   options={Constants.DEFAULT}
                   preselected={details.genres}
                   onAction={(data) => multiselectValidate(data)}/>
      <Input id="movie-overview"
             title="Overview"
             visibility=""
             value={details.description}
             onChange={validate}/>
      <Input id="movie-runtime"
             title="Runtime"
             visibility=""
             value={details.runtime}
             onChange={validate}/>
      <div className="add-movie-actions">
        <button type="button"
                className="reset-button"
                onClick={() => reset()}>RESET
        </button>
        <button type="button"
                className="submit-button"
                onClick={() => submit()}>SUBMIT
        </button>
      </div>
    </form>
  </>
);

MovieEdit.propTypes = {
  modalTitle: PropTypes.string.isRequired,
  closeAction: PropTypes.func.isRequired,
  details: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    title: PropTypes.string,
    release: PropTypes.string,
    poster: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    description: PropTypes.string,
    runtime: PropTypes.number,
  }).isRequired,
};

export default MovieEdit;
