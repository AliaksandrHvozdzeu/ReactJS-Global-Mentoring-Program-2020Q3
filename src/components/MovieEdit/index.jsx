import React from 'react';
import PropTypes from 'prop-types';
import Input from '../Input';
import MultiSelect from '../MultiSelectInput';
import Constants from '../../constants';
import './style.css';
import CloseButton from '../CloseButton';
import Calendar from '../Calendar';

const MovieEdit = ({ modalTitle, closeAction, details }) => (
  <>
    <CloseButton closeAction={closeAction}/>
    <form className="add-movie" onSubmit={() => console.log('form submitted')}>
      <p className="modal-title">{modalTitle}</p>
      <Input id="movie-id"
             title="Movie id"
             visibility=""
             value={details.id}
             onChange={console.log('Validate')}/>
      <Input id="movie-title"
             title="Title"
             visibility=""
             value={details.title}
             onChange={console.log('Validate')}/>
      <Calendar id="movie-release"
                title="Release date"
                value={details.release}
                onChange={console.log('Validate')}/>
      <Input id="movie-url"
             title="Movie url"
             visibility=""
             value={details.poster}
             onChange={console.log('Validate')}/>
      <MultiSelect title="genre"
                   options={Constants.DEFAULT}
                   preselected={details.genres}
                   onAction={(data) => console.log(data)}/>
      <Input id="movie-overview"
             title="Overview"
             visibility=""
             value={details.description}
             onChange={console.log('Validate')}/>
      <Input id="movie-runtime"
             title="Runtime"
             visibility=""
             value={details.runtime}
             onChange={console.log('Validate')}/>
      <div className="add-movie-actions">
        <button className="reset-button"
                onClick={() => console.log('Reset action')}>RESET</button>
        <button className="submit-button"
                onClick={() => console.log('Submit action')}>SUBMIT</button>
      </div>
    </form>
  </>
);

MovieEdit.propTypes = {
  closeAction: PropTypes.func.isRequired,
  details: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    title: PropTypes.string,
    release: PropTypes.string,
    poster: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    description: PropTypes.string,
    runtime: PropTypes.number,
  }).isRequired,
};

export default MovieEdit;
