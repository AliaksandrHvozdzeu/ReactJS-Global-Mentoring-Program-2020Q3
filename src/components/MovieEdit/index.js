import MovieEdit from './js/MovieEdit';

export default MovieEdit;
