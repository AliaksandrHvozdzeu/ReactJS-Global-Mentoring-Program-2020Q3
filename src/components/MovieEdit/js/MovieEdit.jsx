import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { movieActions } from '../../../store/actions';
import Constants from '../../../constants';
import '../css/MovieEdit.css';
import MovieEditForm from './MovieEditForm';

const MovieEdit = ({ modalTitle, closeAction, initialState = {}, onSubmit }) => {

  const availableGenres = useMemo(() => {
    const defaultGenresNames = Constants.DEFAULT.map((genre) => genre.value);
    const uniqueGenres = [...new Set([...initialState.genres, ...defaultGenresNames])].sort();
    return uniqueGenres.map((genre) => ({ label: genre, value: genre }));
  }, [initialState.genres]);

  return (
    <MovieEditForm closeAction={closeAction}
                   modalTitle={modalTitle}
                   initialState={initialState}
                   availableGenres={availableGenres}
                   onSubmit={onSubmit}/>
  );

};

MovieEdit.propTypes = {
  modalTitle: PropTypes.string.isRequired,
  closeAction: PropTypes.func.isRequired,
  initialState: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

MovieEdit.defaultProps = {
  initialState: {},
};

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (data) => dispatch(movieActions.editMovie(data)),
});

const mapStateToProps = (state) => ({
  genres: state.movies.genres.filter((genre) => genre !== Constants.ALL_GENRES),
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieEdit);

