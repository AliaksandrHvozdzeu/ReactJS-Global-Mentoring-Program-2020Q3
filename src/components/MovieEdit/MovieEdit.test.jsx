import React from "react";
import rerender from "react-test-renderer";
import MovieEdit from "./js/MovieEdit";
import { jest } from "@jest/globals";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

describe("when MovieEdit", () => {

  test("then snapshot created", () => {
    const closeAction = jest.fn();
    const modalTitle = "EDIT FORM";
    const initialState = {
      "id": 337167,
      "title": "Fifty Shades Freed",
      "tagline": "Don't miss the climax",
      "vote_average": 6.1,
      "vote_count": 1195,
      "release_date": "2018-02-07",
      "poster_path": "https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg",
      "overview": "Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.",
      "budget": 55000000,
      "revenue": 136906000,
      "genres": [
        "Drama",
        "Romance"
      ],
      "runtime": 106
  };
    const onSubmit = jest.fn();
    const initialStore = {
      movies: {
        "genres": [
          "Drama",
          "Romance"
        ],
      }
    };
    const store = configureStore([])(initialStore);
    store.dispatch = jest.fn();
    const component = rerender.create(<Provider store={store}>
      <MovieEdit closeAction={closeAction}
                 modalTitle={modalTitle}
                 initialState={initialState}
                 availableGenres={[]}
                 onSubmit={onSubmit} /></Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
