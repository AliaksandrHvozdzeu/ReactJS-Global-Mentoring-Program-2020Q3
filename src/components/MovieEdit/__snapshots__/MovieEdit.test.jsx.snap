// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when MovieEdit then snapshot created 1`] = `
Array [
  <button
    aria-label="close"
    className="close-button"
    onClick={[MockFunction]}
    type="button"
  />,
  <form
    className="add-movie"
    onSubmit={[Function]}
  >
    <p
      className="modal-title"
    >
      EDIT FORM
    </p>
    <label
      htmlFor="movie-id"
    >
      Movie id
      <p
        className="text"
        id="movie-id"
      >
        337167
      </p>
    </label>
    <label
      className=""
      htmlFor="movie-title"
    >
      Title
      <input
        className="input"
        id="movie-title"
        name="title"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="text"
        value="Fifty Shades Freed"
      />
    </label>
    <label
      htmlFor="movie-release"
    >
      Release date
      <input
        className="input"
        id="movie-release"
        name="release_date"
        onBlur={[Function]}
        onChange={[Function]}
        type="date"
        value="2018-02-07"
      />
    </label>
    <label
      className=""
      htmlFor="movie-url"
    >
      Movie url
      <input
        className="input"
        id="movie-url"
        name="poster_path"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="url"
        value="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
      />
    </label>
    <label
      className="genres"
      htmlFor="genres"
      id="malty"
    >
      genre
      <div
        className="go2139093995 multi-select"
      >
        <div
          aria-expanded={false}
          aria-labelledby="malty"
          aria-readonly={true}
          className="go2646822163 dropdown-container"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={0}
        >
          <div
            className="go2642161244 dropdown-heading"
            onClick={[Function]}
          >
            <div
              className="dropdown-heading-value"
            >
              <span>
                Drama, Romance
              </span>
            </div>
            <img
              alt="arrow_down"
              className="arrow"
              src="/images/arrow_down.png"
            />
          </div>
        </div>
      </div>
    </label>
    <label
      className=""
      htmlFor="movie-overview"
    >
      Overview
      <input
        className="input"
        id="movie-overview"
        name="overview"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="text"
        value="Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins."
      />
    </label>
    <label
      className=""
      htmlFor="movie-runtime"
    >
      Runtime
      <input
        className="input"
        id="movie-runtime"
        name="runtime"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        type="number"
        value={106}
      />
    </label>
    <div
      className="add-movie-actions"
    >
      <button
        className="reset-button"
        onClick={[Function]}
        type="button"
      >
        RESET
      </button>
      <button
        className="submit-button"
        onClick={[Function]}
        type="button"
      >
        SAVE
      </button>
    </div>
  </form>,
]
`;
